src/pybind/mgr/ansible_orchestrator/module.py

```diff
@@ -0,0 +1,364 @@
+        """Execute the playbook with the provided params.
+        """
+
+        self.pb_execution.launch()
+
+    def update_result(self):
+        """Output of the read operation
+
+        The result of the playbook execution can be customized through the
+        function provided as 'process_output' attribute
+
+        @return string: Result of the operation formatted if it is possible
+        """
+
+        processed_result = []
+
+        if self._is_complete:
+            raw_result = self.pb_execution.get_result(self.event_filter)
+
+            if self.process_output:
+                processed_result = self.process_output(
+                                            raw_result,
+                                            self.ar_client,
+                                            self.pb_execution.play_uuid)
+            else:
+                processed_result = raw_result
+
+        self._result = processed_result
+
+
+class AnsibleChangeOperation(orchestrator.WriteCompletion):
+    """Operations that changes the "cluster" state
+
+    Modifications/Changes (writes) are a two-phase thing, firstly execute
+    the playbook that is going to change elements in the Ceph Cluster.
+    When the playbook finishes execution (independently of the result),
+    the modification/change operation has finished.
+    """
+    def __init__(self):
+        super(AnsibleChangeOperation, self).__init__()
+
+        self.error = False
+    @property
+    def status(self):
+        """Return the status code of the operation
+        """
+        #TODO
+        return 0
+
+    @property
+    def is_persistent(self):
+        """
+        Has the operation updated the orchestrator's configuration
+        persistently?  Typically this would indicate that an update
+        had been written to a manifest, but that the update
+        had not necessarily been pushed out to the cluster.
+
+        In the case of Ansible is always False.
+        because a initiated playbook execution will need always to be
+        relaunched if it fails.
+        """
+
+        return False
+
+    @property
+    def is_effective(self):
+        """Has the operation taken effect on the cluster?
+        For example, if we were adding a service, has it come up and appeared
+        in Ceph's cluster maps?
+
+        In the case of Ansible, this will be True if the playbooks has been
+        executed succesfully.
+
+        @return Boolean: if the playbook has been executed succesfully
+        """
+
+        return self.status == ExecutionStatusCode.SUCCESS
+
+    @property
+    def is_errored(self):
+        return self.error
+
+    @property
+    def is_complete(self):
+        return self.is_errored or (self.is_persistent and self.is_effective)
+
+
+class Module(MgrModule, orchestrator.Orchestrator):
+    """An Orchestrator that an external Ansible runner service to perform
+    operations
+    """
+
+    OPTIONS = [
+        {'name': 'server_addr'},
```

Comment:
I'd suggest merging the addr+port settings into a single URL setting.

Reply:
Why?

Reply:
So that the server's URL is set atomically (i.e. port and hostname together) -- if changing the server's address and port, you don't want to go through an intermediate stage where it's trying to talk to the wrong port on the right hostname, or vice versa.

Reply:
Ok... but in the orchestrator these values are not effective when you change them, only when you disable/enable the node ( unless we change implementation to allow "hot" change of config variables)

Reply:
Yes, I'm assuming that you would at some point want to improve the connection/authentication stuff so that an authentication error or a config change didn't require a ceph-mgr restart.

Reply:
Disabling or enabling a module restarts ceph-mgr.

Reply:
8-| ... i didn't realize of that... in fact the service is not restarted, the pid of the binary continues being the same ... but internally as you say a restart is executed .... 
Does not seem very healthy the fact of changing and make effective a setting in one module, force the restart of other modules ... now i understand your comment:
> you would at some point want to improve the connection/authentication stuff so that an authentication error or a config change didn't require a ceph-mgr restart

But this is a problem that affects all the modules... 
So what i think what we need is basically a method that refreshes configuration and can be called from CLI.

I propose:
- To add in the MgrModule base class  a method "refresh_config" to be overwritten by modules: 
In this method the module must read all the settings and apply the changes detected.
- To implement in the Ansible Orchestrator this method:
- Add in the Orchestrator_CLI a new command to call the "refresh_config" in the backend orchestrator

If you are agree... i can do this... i think that this is really 'adding value' ... more than join/not join together two different settings.

Reply:
Firstly, I still think that you should store your URL as a URL.  Your web browser doesn't have different text boxes for the hostname vs. port, and neither should your settings.  Trust me on this.  The way to specify the destination for an HTTP connection is to have a URL setting, this isn't controversial.

While some modules would benefit from a notification on changes (and that could be implemented pretty easily from PyModuleRegistry::handle_config calling through to ActivePyModules::notify_all), you don't need it here.  Because you're a client rather than a server, you can just look at the configured URL each time you make a request.  If it's different from your established client session, just throw away your session and open a new one.

Reply:
Done. server + port is now one setting: server_url